### DISCUSSION TITLE
When dual stack records, the IPV4 result is returned preferentially

### DISCUSSION THREAD

**Discussion by kkkgo** [DNSCrypt/dnscrypt-proxy#2229] [Ideas]
It is well known that China has serious DNS pollution and IP blocking. On the other hand, most home broadband in China now has IPV6 dual stack. DNSCRYPT is used with proxy software in most cases in China, and proxy software usually only supports connections from IPV4 addresses, so for most blocked websites, returning IPV6 addresses will cause trouble. The option "block_ipv6=true" solves this problem nicely. But there is a situation, depending on the censorship of the national firewall, some IPV6 only websites can be directly linked, and these websites do not have IPV4 addresses, which makes me unable to access these websites.
Therefore, I want to add an option, such as "prefer_ipv4=true", when returning dual stack results, filter out IPv6 results, and return IPv6 results when the result is only ipv6.

**Comment by kkkgo** [comment]
I found a similar function in another project, I hope it can provide a useful reference：
https://github.com/IrineSistiana/mosdns/blob/main/plugin/executable/dual_selector/dual_selector.go

**Reply by kkkgo** [reply]
According to the project's wiki description, _prefer_ipv4 can be used to allow downstream clients to prefer IPv4 when supporting dual stack. _prefer_ipv4 will automatically detect whether the domain name is dual-stack, and then only block AAAA requests for dual-stack domain names. Compared with the scheme of forcibly disabling all AAAA requests, this scheme does not affect the resolution of pure IPv6 domain names.
https://irine-sistiana.gitbook.io/mosdns-wiki/mosdns/cha-jian-ji-qi-can-shu#preferipv46-you-xian-shi-yong-ipv46

**Comment by 57382** [comment]
there is another possible situation:
The website has IPv4 and IPv6 addresses. IPv4 addresses cannot be accessed. IPv6 addresses can be accessed.
It seems that I also need an option: prefer_ipv6=true
Prioritize returning IPv6 addresses. I also need your V4 option. It would be great if DNSCrypt could intelligently switch between these two modes, and the address that can be accessed normally takes priority, no matter it is V4 or V6.
